shops/views.py

```diff
@@ -232,7 +234,50 @@ def upload_data(request):
     return render(request, 'shops/upload_data.html', {'form': form})
 
 
+class SummaryReportView(generic.TemplateView):
+    template_name = 'shops/summary_report.html'
+
+    def get_context_data(self, *args, **kwargs):
+        # For each fruit, for each 4-week period
+        start_date = date(2021, 1, 4)
+        sale_data = models.WeeklySale.objects.filter(
+            weekly_shop_summary__date__gte=start_date
+        ).annotate(
+            revenue=F('units_sold')*F('price_per_unit')
+        )
+
+        sales_summary = sale_data.annotate(
+            period_number=(functions.ExtractWeek('weekly_shop_summary__date')+3)/4,
+            keep=(functions.ExtractWeek('weekly_shop_summary__date')+3)%4
+        ).annotate(
+            total_revenue=Window(
+                expression=Sum('revenue'),
+                partition_by=[
+                    F('fruit__name'), 
+                    F('weekly_shop_summary__date__year'),
```

Comment:
Do you need to partition on year when you're also partitioning it on period?

Reply:
If the data spans over 1 year, there will be repeating period numbers which must be separated.

Reply:
Ah, okay. I'd assumed that the period included the year, but it only includes the week within a year. All clear now, thanks